sycl/include/CL/sycl/handler.hpp

```diff
@@ -366,6 +367,109 @@ class __SYCL_EXPORT handler {
     return true;
   }
 
+  /// Handles some special cases of the copy operation from one accessor
+  /// to another accessor. Returns true if the copy is handled here.
+  ///
+  /// \param Src is a source SYCL accessor.
+  /// \param Dst is a destination SYCL accessor.
+  // TODO: support atomic accessor in Src or/and Dst.
+  template<
+      typename TSrc, int DimSrc, access::mode ModeSrc, access::target TargetSrc,
+      typename TDst, int DimDst, access::mode ModeDst, access::target TargetDst,
+      access::placeholder IsPHSrc, access::placeholder IsPHDst>
+  detail::enable_if_t<(DimSrc > 0) && (DimDst > 0), bool>
+  copyAccToAccHelper(
+      accessor<TSrc, DimSrc, ModeSrc, TargetSrc, IsPHSrc> Src,
+      accessor<TDst, DimDst, ModeDst, TargetDst, IsPHDst> Dst) {
+    if (!MIsHost &&
+        IsCopyingRectRegionAvailable(Src.get_range(), Dst.get_range()))
+      return false;
+
+    range<1> LinearizedRange(Src.get_count());
+    parallel_for<class __copyAcc2Acc<TSrc, DimSrc, ModeSrc, TargetSrc,
+                                     TDst, DimDst, ModeDst, TargetDst,
+                                     IsPHSrc, IsPHDst>>
+                                     (LinearizedRange, [=](id<1> Id) {
+      size_t Index = Id[0];
+      id<DimSrc> SrcIndex = getDelinearizedIndex(Src.get_range(), Index);
+      id<DimDst> DstIndex = getDelinearizedIndex(Dst.get_range(), Index);
+      Dst[DstIndex] = Src[SrcIndex];
+    });
+    return true;
+  }
+
+  template <typename T, int Dim, access::mode Mode, access::target Target,
+            access::placeholder IsPH>
+  detail::enable_if_t<Dim == 0 && Mode == access::mode::atomic, T>
+  readFromFirstAccElement(accessor<T, Dim, Mode, Target, IsPH> Src) const {
+    auto AtomicSrc = (atomic<T, access::address_space::global_space>)Src;
```

Comment:
Why do we need to explicitly convert here? The spec says that operator[] of accessor with access mode == atomic returns object of atomic class.

Reply:
Operator [] is available only when Dim > 0:
97 /* Available only when: accessMode == access::mode::atomic && dimensions >
98 0 */
99 atomic<dataT, access::address_space::global_space> operator[](
100 id<dimensions> index) const;

The cast is needed when Dim==0